plugin/db/mysql/mysql.go

```diff
@@ -938,6 +950,104 @@ func (driver *Driver) Restore(ctx context.Context, sc *bufio.Scanner) (err error
 	return nil
 }
 
+// Ghost table names are generated by appending "_ghost" to the original ones.
+// Constraint names are generated by appending a timestamp (to second) to the original ones.
+// According to https://dev.mysql.com/doc/refman/8.0/en/identifier-length.html, the max length of a
+// constraint name is 64, so we should truncate the original constraint name part to fit in 64 characters.
+func rewriteToGhostTableForRestore(stmt, timestamp string) string {
+	reInsert := regexp.MustCompile(`INSERT INTO ` +
+		// Capture group for optional db prefix like "`dbPrefix`."
+		// the ?: means DO NOT capture this group
+		// the last ? means the whole db prefix is optional
+		"(?:`([^`]+)`\\.)?" +
+		// Capture group for the table name, leaving back quote ` outside for convenience
+		"`([^`]+)` " +
+		// Capture group for values to the end of the line
+		"VALUES (.*)")
+
+	reCreate := regexp.MustCompile("CREATE TABLE " +
+		// Capture group for the table name, leaving back quote ` outside for convenience
+		"`([^`]+)` " +
+		// Capture group for the rest of the contents
+		// Note that there will be multiple lines, so using [\s\S]+ to match them all including line breaks
+		`([\s\S]+)`)
+
+	// matchesInsert would be like:
+	// with dbPrefix: []string{"INSERT INTO `dbPrefix`.`table` VALUES (1, 2, 3);", "dbPrefix", "table", "(1, 2, 3);"}
+	// empty dbPrefix: []string{"INSERT INTO `table` VALUES (1, 2, 3);", "", "table", "(1, 2, 3);"}
+	matchesInsert := reInsert.FindStringSubmatch(stmt)
+
+	// matchesCreate would be like:
+	// []string{"CREATE TABLE `table` (", "table", "("}
+	matchesCreate := reCreate.FindStringSubmatch(stmt)
+
+	var stmtWithGhostTables string
+	if matchesInsert != nil {
+		dbPrefix := ""
+		if matchesInsert[1] != "" {
+			dbPrefix = fmt.Sprintf("`%s`.", matchesInsert[1])
+		}
+		ghostTable := fmt.Sprintf("%s_ghost", matchesInsert[2])
+		values := matchesInsert[3]
+		stmtWithGhostTables = fmt.Sprintf("INSERT INTO %s`%s` VALUES %s", dbPrefix, ghostTable, values)
+	} else if matchesCreate != nil {
+		// In the create table case, we must remove the constraint name from the SQL
+		// (foreign key/check constraints), or a duplicate constraint error will be raised by MySQL.
+		ghostTable := fmt.Sprintf("%s_ghost", matchesCreate[1])
+		tableDef := matchesCreate[2]
+
+		// matches CONSTRAINT names, only match in one line
```

Comment:
Do we need to restore the constraints for creating the gh-ost table? Removing the constraint can save some complexity.

Reply:
If we remove the constraint, users will find constraints are missing after a restore. I suppose it is not expected?

Reply:
If I understand the correctly, these are the constraints on the gh-ost table which should be ghost. In the end of the workflow, we'll swap gh-ost table and primary table where primary table still has the constraints.

Reply:
Here's the basic logic for resolving constraint name conflict:
1. the original table has a constraint named `chk_a_1`
2. we create the ghost table, with a constraint named `chk_a_1_202204142007`
3. after restore, we swap the original and ghost table, then the table will have a constraint with name `chk_a_1_202204142007`